### PR TITLE
Add subpackages module to skylib to support new bazel native.subpackages

### PR DESCRIPTION
**Pull request by kkress**
This is a stacked PR on top of https://github.com/bazelbuild/bazel-skylib/pull/347 (commit #1), this should not be merged until that PR is accepted and merged.

This PR adds some simple utility functions for the new `native.subpackages` loading phase function added to bazel.  It allows callers to check for support of the function in bazel before calling and provides some useful utility.

### CODE REVIEW

**Comment by RoyMoto** [comment]
I fournd typo.
_SUBACKAGES_SUPPORTED

File: lib/subpackages.bzl [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    return "//%s/%s" % (native.package_name(), relative_path)
+
+def _exists(relative_path):
+    """Checks to see if relative_path is a direct subpackage of the current package.
+
+    Example:
+
+    BUILD
+    foo/BUILD
+    foo/sub/BUILD
```
Comment (tetromino) [inline lib/subpackages.bzl:74]: SImilarly to above: please indent by 4 additional spaces

File: lib/subpackages.bzl [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    If the current package is 'BUILD':
+    subpackages.exists("foo") == True
+    subpackages.exists("foo/sub") == False
+    subpackages.exists("bar") == False
```
Comment (tetromino) [inline lib/subpackages.bzl:79]: SImilarly to above: please indent by 4 additional spaces

File: lib/subpackages.bzl [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+def _exists(relative_path):
+    """Checks to see if relative_path is a direct subpackage of the current package.
+
+    Example:
+
+    BUILD
+    foo/BUILD
+    foo/sub/BUILD
+
+    If the current package is 'BUILD':
```
Comment (tetromino) [inline lib/subpackages.bzl:76]: Nit, similarly to above: "If the current package is `//`" or "In the context of the top-level BUILD file"

**Review by tetromino — CHANGES_REQUESTED** [review]

File: lib/subpackages.bzl [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    The returned list contains all subpackages, but not subpackages of subpackages.
+
+    Example:
+    Assuming the following BUILD files exist:
+
+    BUILD
+    foo/BUILD
+    foo/sub/BUILD
+    bar/BUILD
+    baz/deep/dir/BUILD
```
[the remark this answers is not in the document]
Comment (kkress) [reply lib/subpackages.bzl]: Done.

File: lib/subpackages.bzl [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    If the current package is 'BUILD' all() will return ['//foo', '//bar',
```
[the remark this answers is not in the document]
Comment (kkress) [reply lib/subpackages.bzl]: Done.

File: lib/subpackages.bzl [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylib module containing common functions for working with native.subpackages()
+"""
+_SUBACKAGES_SUPPORTED = "subpackages" in dir(native)
```
[the remark this answers is not in the document]
Comment (kkress) [reply lib/subpackages.bzl]: Done.

File: lib/subpackages.bzl [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+def _supported():
+    return _SUBACKAGES_SUPPORTED
+
+def _check_supported():
+    if not _SUBACKAGES_SUPPORTED:
+        fail("native.subpackages not supported.")
```
[the remark this answers is not in the document]
Comment (kkress) [reply lib/subpackages.bzl:24]: Done.

File: lib/subpackages.bzl (hunk 3 above)
Comment (kkress) [reply lib/subpackages.bzl:76]: Done.

**Comment by kkress** [comment]
OK I think that did it, sorry my git/github foo is not great so there are a bunch of commits :/

File: lib/subpackages.bzl (hunk 1 above)
Comment (kkress) [reply lib/subpackages.bzl]: Done.

File: tests/subpackages_tests.bzl [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        subpackages.all(fully_qualified = False),
+    )
+
+    # test exclusion
+    loadingtest.asserts(
+        env,
+        "all_w_exclude",
+        filtered_pkgs,
+        subpackages.all(exclude = ["diff_test", "run_binary", "native_binary"], fully_qualified = False),
+    )
```
[the remark this answers is not in the document]
Comment (kkress) [reply tests/subpackages_tests.bzl]: Fixed to assert order.

**Review by tetromino — APPROVED** [review]
Looks good. I had to fix the tests and the BUILD file due to merge conflict.

**Pull request merged**